translator/opencensus/oc_to_internal.go

```diff
@@ -0,0 +1,405 @@
+	dest.SetSpanID(data.SpanIDFromBytes(src.SpanId))
+	dest.SetTraceState(ocTraceStateToInternal(src.Tracestate))
+	dest.SetParentSpanID(data.SpanIDFromBytes(src.ParentSpanId))
+	dest.SetName(truncableStringToStr(src.Name))
+	dest.SetKind(ocSpanKindToInternal(src.Kind, src.Attributes))
+	dest.SetStartTime(timestampToUnixnano(src.StartTime))
+	dest.SetEndTime(timestampToUnixnano(src.EndTime))
+	dest.SetAttributes(ocAttrsToInternal(src.Attributes))
+	dest.SetEvents(events)
+	dest.SetDroppedEventsCount(droppedEventCount)
+	dest.SetLinks(links)
+	dest.SetDroppedLinksCount(droppedLinkCount)
+	dest.SetStatus(ocStatusToInternal(src.Status))
+}
+
+func ocStatusToInternal(ocStatus *octrace.Status) data.SpanStatus {
+	if ocStatus == nil {
+		return data.SpanStatus{}
+	}
+	return data.NewSpanStatus(data.StatusCode(ocStatus.Code), ocStatus.Message)
+}
+
+// Convert tracestate to W3C format. See the https://w3c.github.io/trace-context/
+func ocTraceStateToInternal(ocTracestate *octrace.Span_Tracestate) data.TraceState {
+	if ocTracestate == nil {
+		return ""
+	}
+	var sb strings.Builder
+	for i, entry := range ocTracestate.Entries {
+		if i > 0 {
+			sb.WriteString(",")
+		}
+		sb.WriteString(strings.Join([]string{entry.Key, entry.Value}, "="))
+	}
+	return data.TraceState(sb.String())
+}
+
+func ocAttrsToInternal(ocAttrs *octrace.Span_Attributes) data.Attributes {
+	if ocAttrs == nil {
+		return data.Attributes{}
+	}
+
+	attrCount := len(ocAttrs.AttributeMap)
+
+	attrMap := make(data.AttributesMap, attrCount)
+	values := data.NewAttributeValueSlice(attrCount)
+	i := 0
+	for key, ocAttr := range ocAttrs.AttributeMap {
+		switch attribValue := ocAttr.Value.(type) {
+		case *octrace.AttributeValue_StringValue:
+			values[i].SetString(truncableStringToStr(attribValue.StringValue))
+
+		case *octrace.AttributeValue_IntValue:
+			values[i].SetInt(attribValue.IntValue)
+
+		case *octrace.AttributeValue_BoolValue:
+			values[i].SetBool(attribValue.BoolValue)
+
+		case *octrace.AttributeValue_DoubleValue:
+			values[i].SetDouble(attribValue.DoubleValue)
+
+		default:
+			str := "<Unknown OpenCensus Attribute>"
+			values[i].SetString(str)
+		}
+		attrMap[key] = values[i]
+		i++
+	}
+	droppedCount := uint32(ocAttrs.DroppedAttributesCount)
+	return data.NewAttributes(attrMap, droppedCount)
+}
+
+func ocSpanKindToInternal(ocKind octrace.Span_SpanKind, ocAttrs *octrace.Span_Attributes) data.SpanKind {
+	switch ocKind {
+	case octrace.Span_SERVER:
+		return data.SpanKindSERVER
+
+	case octrace.Span_CLIENT:
+		return data.SpanKindCLIENT
+
+	case octrace.Span_SPAN_KIND_UNSPECIFIED:
+		// Span kind field is unspecified, check if TagSpanKind attribute is set.
+		// This can happen if span kind had no equivalent in OC, so we could represent it in
+		// the SpanKind. In that case the span kind may be a special attribute TagSpanKind.
+		if ocAttrs != nil {
+			kindAttr := ocAttrs.AttributeMap[tracetranslator.TagSpanKind]
+			if kindAttr != nil {
```

Comment:
It's more idiomatic to use the `ok` flag:
```
if kindAttr, ok := ocAttrs.AttributeMap[tracetranslator.TagSpanKind]; ok {
  ...
}
```
Unless you want to handle attribute with explicitly set nil value here (then you can probably just change condition to `ok && kindAttr != nil` to make it more explicit?

Reply:
I don't know about "more", but relying on zero values is pretty idiomatic. See https://golang.org/doc/effective_go.html#maps

"comma ok" idiom is _needed_ when you care about distinguishing missing and zero value. In this case we don't care and using "ok" adds code and benefits are arguable.